Union and intersection types are not causing checking the kind of type parameters

## Compiler version
3.2.1

## Minimized code

<!--
This code should be self contained, compilable (with possible failures) and as small as possible.

Ideally, we should be able to just copy this code in a file and run `scalac` (and maybe `scala`) to reproduce the issue.
-->
https://scastie.scala-lang.org/VEozYp5OQlmLyTb0hn3KQw

## Output (click arrow to expand)
<details>

```scala
java.lang.AssertionError: assertion failed: R &amp; Playground.Foo / TypeParamRef(R) &amp; HKTypeLambda(List(A), List(TypeBounds(TypeRef(ThisType(TypeRef(NoPrefix,module class scala)),class Nothing),TypeRef(ThisType(TypeRef(NoPrefix,module class scala)),class Any))), AppliedType(TypeRef(ThisType(TypeRef(ThisType(TypeRef(NoPrefix,module class &lt;empty&gt;)),module class Playground$)),trait Foo),List(TypeParamRef(A))))
	at scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
	at dotty.tools.dotc.core.Types$AndType$.apply(Types.scala:3203)
	at dotty.tools.dotc.core.Types$AndType$.make(Types.scala:3238)
	at dotty.tools.dotc.core.Types$AndType.derivedAndType(Types.scala:3184)
	at dotty.tools.dotc.core.Types$ApproximatingTypeMap.derivedAndType(Types.scala:5961)
	at dotty.tools.dotc.core.Types$TypeMap.mapOver(Types.scala:5689)
	at dotty.tools.dotc.core.TypeOps$AsSeenFromMap.apply(TypeOps.scala:109)
	at dotty.tools.dotc.core.Types$TypeMap.op$proxy16$1(Types.scala:5590)
	at dotty.tools.dotc.core.Types$TypeMap.mapArgs(Types.scala:5590)
	at dotty.tools.dotc.core.Types$TypeMap.mapOver(Types.scala:5627)
	at dotty.tools.dotc.core.TypeOps$AsSeenFromMap.apply(TypeOps.scala:109)
	at dotty.tools.dotc.core.Types$TypeMap.mapOverLambda(Types.scala:5603)
	at dotty.tools.dotc.core.TypeOps$AsSeenFromMap.apply(TypeOps.scala:103)
	at dotty.tools.dotc.core.Types$TypeMap.mapOverLambda(Types.scala:5603)
	at dotty.tools.dotc.core.TypeOps$AsSeenFromMap.apply(TypeOps.scala:103)
	at dotty.tools.dotc.core.Types$TypeMap.mapOverLambda(Types.scala:5603)
	at dotty.tools.dotc.core.TypeOps$AsSeenFromMap.apply(TypeOps.scala:103)
	at dotty.tools.dotc.core.TypeOps$.asSeenFrom(TypeOps.scala:51)
	at dotty.tools.dotc.core.Types$Type.asSeenFrom(Types.scala:1036)
	at dotty.tools.dotc.core.Denotations$SingleDenotation.derived$1(Denotations.scala:1099)
	at dotty.tools.dotc.core.Denotations$SingleDenotation.computeAsSeenFrom(Denotations.scala:1125)
	at dotty.tools.dotc.core.Denotations$SingleDenotation.computeAsSeenFrom(Denotations.scala:1079)
	at dotty.tools.dotc.core.Denotations$PreDenotation.asSeenFrom(Denotations.scala:134)
	at dotty.tools.dotc.core.SymDenotations$ClassDenotation.findMember(SymDenotations.scala:2069)
	at dotty.tools.dotc.core.Types$Type.go$1(Types.scala:686)
	at dotty.tools.dotc.core.Types$Type.findMember(Types.scala:879)
	at dotty.tools.dotc.core.Types$Type.memberBasedOnFlags(Types.scala:669)
	at dotty.tools.dotc.core.Types$Type.member(Types.scala:653)
	at dotty.tools.dotc.typer.ProtoTypes$SelectionProto.liftedTree1$1(ProtoTypes.scala:195)
	at dotty.tools.dotc.typer.ProtoTypes$SelectionProto.isMatchedBy(ProtoTypes.scala:217)
	at dotty.tools.dotc.core.TypeComparer.isMatchedByProto(TypeComparer.scala:1917)
	at dotty.tools.dotc.core.TypeComparer.firstTry$1(TypeComparer.scala:314)
	at dotty.tools.dotc.core.TypeComparer.recur(TypeComparer.scala:1336)
	at dotty.tools.dotc.core.TypeComparer.isSubType(TypeComparer.scala:194)
	at dotty.tools.dotc.core.TypeComparer.isSubType(TypeComparer.scala:204)
	at dotty.tools.dotc.core.TypeComparer.topLevelSubType(TypeComparer.scala:128)
	at dotty.tools.dotc.core.TypeComparer.testSubType(TypeComparer.scala:144)
	at dotty.tools.dotc.core.TypeComparer$.testSubType(TypeComparer.scala:2767)
	at dotty.tools.dotc.typer.Typer.adaptNoArgsOther$1(Typer.scala:3725)
	at dotty.tools.dotc.typer.Typer.adaptNoArgs$1(Typer.scala:3822)
	at dotty.tools.dotc.typer.Typer.adapt1(Typer.scala:4038)
	at dotty.tools.dotc.typer.Typer.adapt(Typer.scala:3368)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2987)
```
</details>

--- a/compiler/src/dotty/tools/dotc/core/Types.scala
+++ b/compiler/src/dotty/tools/dotc/core/Types.scala
@@ -3509,7 +3509,7 @@ object Types {
   def expectValueTypeOrWildcard(tp: Type, where: => String)(using Context): Unit =
     if !tp.isValueTypeOrWildcard then
       assert(!ctx.isAfterTyper, where) // we check correct kinds at PostTyper
-      throw TypeError(em"$tp is not a value type, cannot be used in $where")
+      throw TypeError(em"$tp is not a value type, cannot be used $where")
 
   /** An extractor object to pattern match against a nullable union.
    *  e.g.

--- a/compiler/src/dotty/tools/dotc/core/Types.scala
+++ b/compiler/src/dotty/tools/dotc/core/Types.scala
@@ -1071,7 +1071,7 @@ object Types {
      *  @param relaxedCheck   if true type `Null` becomes a subtype of non-primitive value types in TypeComparer.
      *  @param matchLoosely   if true the types `=> T` and `()T` are seen as overriding each other.
      *  @param checkClassInfo if true we check that ClassInfos are within bounds of abstract types
-     * 
+     *
      *  @param isSubType      a function used for checking subtype relationships.
      */
     final def overrides(that: Type, relaxedCheck: Boolean, matchLoosely: => Boolean, checkClassInfo: Boolean = true,
@@ -3322,11 +3322,11 @@ object Types {
   final class CachedAndType(tp1: Type, tp2: Type) extends AndType(tp1, tp2)
 
   object AndType {
-    def apply(tp1: Type, tp2: Type)(using Context): AndType = {
-      assert(tp1.isValueTypeOrWildcard &&
-             tp2.isValueTypeOrWildcard, i"$tp1 & $tp2 / " + s"$tp1 & $tp2")
+    def apply(tp1: Type, tp2: Type)(using Context): AndType =
+      def where = i"in intersection $tp1 & $tp2"
+      expectValueTypeOrWildcard(tp1, where)
+      expectValueTypeOrWildcard(tp2, where)
       unchecked(tp1, tp2)
-    }
 
     def balanced(tp1: Type, tp2: Type)(using Context): AndType =
       tp1 match
@@ -3366,7 +3366,7 @@ object Types {
       TypeComparer.liftIfHK(tp1, tp2, AndType.make(_, _, checkValid = false), makeHk, _ | _)
   }
 
-  abstract case class OrType(tp1: Type, tp2: Type) extends AndOrType {
+  abstract case class OrType protected(tp1: Type, tp2: Type) extends AndOrType {
     def isAnd: Boolean = false
     def isSoft: Boolean
     private var myBaseClassesPeriod: Period = Nowhere
@@ -3398,9 +3398,6 @@ object Types {
           myFactorCount = tp1.orFactorCount(soft) + tp2.orFactorCount(soft)
         myFactorCount
       else 1
-
-    assert(tp1.isValueTypeOrWildcard &&
-           tp2.isValueTypeOrWildcard, s"$tp1 $tp2")
 
     private var myJoin: Type = _
     private var myJoinPeriod: Period = Nowhere
@@ -3476,6 +3473,9 @@ object Types {
   object OrType {
 
     def apply(tp1: Type, tp2: Type, soft: Boolean)(using Context): OrType = {
+      def where = i"in union $tp1 | $tp2"
+      expectValueTypeOrWildcard(tp1, where)
+      expectValueTypeOrWildcard(tp2, where)
       assertUnerased()
       unique(new CachedOrType(tp1, tp2, soft))
     }
@@ -3505,6 +3505,11 @@ object Types {
     def makeHk(tp1: Type, tp2: Type)(using Context): Type =
       TypeComparer.liftIfHK(tp1, tp2, OrType(_, _, soft = true), makeHk, _ & _)
   }
+
+  def expectValueTypeOrWildcard(tp: Type, where: => String)(using Context): Unit =
+    if !tp.isValueTypeOrWildcard then
+      assert(!ctx.isAfterTyper, where) // we check correct kinds at PostTyper
+      throw TypeError(em"$tp is not a value type, cannot be used in $where")
 
   /** An extractor object to pattern match against a nullable union.
    *  e.g.

--- a/compiler/test/dotty/tools/repl/TabcompleteTests.scala
+++ b/compiler/test/dotty/tools/repl/TabcompleteTests.scala
@@ -235,6 +235,6 @@ class TabcompleteTests extends ReplTest {
   }
 
   @Test def i9334 = initially {
-    assertEquals(Nil, tabComplete("class Foo[T]; classOf[Foo]."))
+    assert(tabComplete("class Foo[T]; classOf[Foo].").contains("getName"))
   }
 }

--- a/tests/neg/i16696.check
+++ b/tests/neg/i16696.check
@@ -1,0 +1,12 @@
+-- Error: tests/neg/i16696.scala:7:29 ----------------------------------------------------------------------------------
+7 |  val boom1 = BoxMaker[Some].make1 // error
+  |                             ^
+  |                             Some is not a value type, cannot be used in in intersection Some & Int
+-- Error: tests/neg/i16696.scala:8:29 ----------------------------------------------------------------------------------
+8 |  val boom2 = BoxMaker[Some].make2 // error
+  |                             ^
+  |                             Some is not a value type, cannot be used in in union Some | Int
+-- Error: tests/neg/i16696.scala:20:27 ---------------------------------------------------------------------------------
+20 |  val boom = BoxMaker[Foo].make(_.foo) // error
+   |                           ^
+   |                           test2.Foo is not a value type, cannot be used in in intersection R & test2.Foo

--- a/tests/neg/i16696.check
+++ b/tests/neg/i16696.check
@@ -1,12 +1,12 @@
 -- Error: tests/neg/i16696.scala:7:29 ----------------------------------------------------------------------------------
 7 |  val boom1 = BoxMaker[Some].make1 // error
   |                             ^
-  |                             Some is not a value type, cannot be used in in intersection Some & Int
+  |                             Some is not a value type, cannot be used in intersection Some & Int
 -- Error: tests/neg/i16696.scala:8:29 ----------------------------------------------------------------------------------
 8 |  val boom2 = BoxMaker[Some].make2 // error
   |                             ^
-  |                             Some is not a value type, cannot be used in in union Some | Int
+  |                             Some is not a value type, cannot be used in union Some | Int
 -- Error: tests/neg/i16696.scala:20:27 ---------------------------------------------------------------------------------
 20 |  val boom = BoxMaker[Foo].make(_.foo) // error
    |                           ^
-   |                           test2.Foo is not a value type, cannot be used in in intersection R & test2.Foo
+   |                           test2.Foo is not a value type, cannot be used in intersection R & test2.Foo

--- a/tests/neg/i16696.scala
+++ b/tests/neg/i16696.scala
@@ -1,0 +1,20 @@
+object test1:
+  class BoxMaker[T] {
+    def make1: T & Int = ???
+    def make2: T | Int = ???
+  }
+
+  val boom1 = BoxMaker[Some].make1 // error
+  val boom2 = BoxMaker[Some].make2 // error
+
+object test2:
+  class Box[R]
+
+  class BoxMaker[T] {
+    def make[R <: T](f: T => Box[R]): Box[R & T] = ???
+  }
+
+  trait Foo[A]{
+    def foo: Box[Foo[Unit]]
+  }
+  val boom = BoxMaker[Foo].make(_.foo) // error

